server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java

```diff
@@ -76,11 +76,19 @@ public ReleasableBytesReference retain() {
         return this;
     }
 
+    /**
+     * Same as {@link #slice} except that the slice is not guaranteed to share the same underlying reference count as this instance.
+     * This method is equivalent to calling {@code .slice(from, length).retain()} but might be more efficient through the avoidance of
+     * retaining unnecessary buffers.
+     */
     public ReleasableBytesReference retainedSlice(int from, int length) {
         if (from == 0 && length() == length) {
             return retain();
         }
         final BytesReference slice = delegate.slice(from, length);
```

Comment:
Funny enough I found this looking into a different optimization that wouldn't build composite bytes references out of releasable ones by unwrapping. While that has its runtime advantages, this change seems much more helpful by saving memory and optimisations like #124947 (if taken a little further that is) could leverage this for even bigger gains.
As it stands, this change actually is quite helpful already when working with `SearchHit` instances sliced out of large buffers at the very least (there might be other cases where it's helpful as well). Also, it avoids a pointless round of ref-counting in the decoder.

Reply:
This could also be taken further by making the logic aware of situations where the slice itself is made up of a subset of the releasable references that the parent instance consists of and only retaining those. Obviously then considerably increasing the complexity of the implementation for a likely limited gain in corner cases.